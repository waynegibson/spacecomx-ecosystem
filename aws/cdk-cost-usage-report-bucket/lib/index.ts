export * from './stacks/cost-usage-report-bucket-stack'
