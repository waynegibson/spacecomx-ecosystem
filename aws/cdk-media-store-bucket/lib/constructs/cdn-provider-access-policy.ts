/* eslint-disable no-new */
import { Construct } from 'constructs'
import * as iam from 'aws-cdk-lib/aws-iam'
import type { Bucket } from 'aws-cdk-lib/aws-s3'

export interface CDNProviderAccessPolicyProps {
  /**
   * The source bucket to use for the CDN provider.
   */
  readonly sourceBucket: Bucket
  /**
   * Create a IAM user and add it to the CDN provider group that has a bucket access policy attached.
   *
   * If this is set to true, a new IAM user will be created and added to the IAM group.
   * If this is set to false, the user will not be created.
   *
   * @default - true
   */
  readonly createIamUser?: boolean
}

/**
 * Creates an IAM group, IAM user and adds a bucket access policy to it.
 */
export class CDNProviderAccessPolicy extends Construct {
  public readonly group: iam.Group
  public readonly user: iam.User | undefined

  constructor (scope: Construct, id: string, props: CDNProviderAccessPolicyProps) {
    super(scope, id)

    const bucket = props.sourceBucket

    this.group = new iam.Group(this, 'Group', {})

    // Create a policy that allows the group to access the bucket
    new iam.Policy(this, 'AccessPolicy', {
      groups: [this.group],
      statements: [
        new iam.PolicyStatement({
          effect: iam.Effect.ALLOW,
          actions: [
            's3:GetObject',
            's3:ListBucket',
            's3:GetBucketLocation'
          ],
          resources: [
            `${bucket.bucketArn}`,
            `${bucket.bucketArn}/*`
          ]
        })
      ]
    })

    // Create a new user if the user is not created
    if (props?.createIamUser === undefined || props.createIamUser) {
      this.user = new iam.User(this, 'User', {
        groups: [this.group]
      })
    }
  }
}
