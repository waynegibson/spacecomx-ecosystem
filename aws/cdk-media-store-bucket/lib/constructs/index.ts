export * from './cdn-provider-access-policy'
