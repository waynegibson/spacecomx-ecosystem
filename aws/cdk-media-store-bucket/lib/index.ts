export * from './stacks/media-store-bucket-stack'
