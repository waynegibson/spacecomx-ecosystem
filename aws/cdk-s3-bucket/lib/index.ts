export * from './s3-bucket'
