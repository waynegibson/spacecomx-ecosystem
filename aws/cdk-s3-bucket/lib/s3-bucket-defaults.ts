import { RemovalPolicy } from 'aws-cdk-lib'
import * as s3 from 'aws-cdk-lib/aws-s3'

/**
 * @internal
 */
export function DefaultS3Props (loggingBucket?: s3.Bucket, lifecycleRules?: s3.LifecycleRule[]): s3.BucketProps {
  return {
    versioned: true,
    enforceSSL: true,
    encryption: s3.BucketEncryption.S3_MANAGED,
    blockPublicAccess: s3.BlockPublicAccess.BLOCK_ALL,
    autoDeleteObjects: false,
    removalPolicy: RemovalPolicy.RETAIN,
    lifecycleRules: lifecycleRules ?? [],
    serverAccessLogsBucket: loggingBucket
  }
}
