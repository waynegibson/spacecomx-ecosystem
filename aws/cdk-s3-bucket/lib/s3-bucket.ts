import { Construct } from 'constructs'
import type * as s3 from 'aws-cdk-lib/aws-s3'
import { Bucket } from 'aws-cdk-lib/aws-s3'
import { consolidateProps } from '@spacecomx/cdk-utils'
import { DefaultS3Props } from './s3-bucket-defaults'

export interface S3BucketProps {
  /**
   * User provided props to override the default props for the S3 Bucket.
   *
   * @default - Default props are used
   */
  readonly bucketProps?: s3.BucketProps
}

export class S3Bucket extends Construct {
  public readonly s3Bucket: s3.Bucket

  constructor(scope: Construct, id: string, props: S3BucketProps = {}) {
    super(scope, id)

    const combinedBucketProps = consolidateProps(DefaultS3Props(), props.bucketProps)
    const s3BucketResponse = new Bucket(scope, 'S3Bucket', combinedBucketProps)

    this.s3Bucket = s3BucketResponse
  }
}
