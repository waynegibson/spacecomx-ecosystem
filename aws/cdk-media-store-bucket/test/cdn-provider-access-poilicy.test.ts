import { Stack } from 'aws-cdk-lib'
import { Template } from 'aws-cdk-lib/assertions'
import * as s3 from 'aws-cdk-lib/aws-s3'
import { CDNProviderAccessPolicy } from '../lib/constructs'

describe('CDN provider access policy construct for the media bucket -', () => {
  test('it should create a default cdn provider access policy.', () => {
    const stack = new Stack()

    const bucket = new s3.Bucket(stack, 'test-bucket', {})

    new CDNProviderAccessPolicy(stack, 'test-cdn-provider-access-policy', {
      sourceBucket: bucket,
    })

    const t = Template.fromStack(stack)

    expect(t).toMatchSnapshot()

    t.resourceCountIs('AWS::IAM::Group', 1)
    t.resourceCountIs('AWS::IAM::User', 1)
    t.resourceCountIs('AWS::IAM::Policy', 1)

    t.hasResourceProperties('AWS::IAM::Policy', {
      PolicyDocument: {
        Statement: [
          {
            Action: [
              's3:GetObject',
              's3:ListBucket',
              's3:GetBucketLocation',
            ],
            Effect: 'Allow',
          },
        ],
      },
    })
  })

  test('it should create a cdn provider access policy without a default IAM user.', () => {
    const stack = new Stack()

    const bucket = new s3.Bucket(stack, 'test-bucket', {})

    new CDNProviderAccessPolicy(stack, 'test-cdn-provider-access-policy', {
      sourceBucket: bucket,
      createIamUser: false,
    })

    const t = Template.fromStack(stack)

    expect(t).toMatchSnapshot()

    t.resourceCountIs('AWS::IAM::Group', 1)
    t.resourceCountIs('AWS::IAM::Policy', 1)
    t.resourceCountIs('AWS::IAM::User', 0)
  })
})
