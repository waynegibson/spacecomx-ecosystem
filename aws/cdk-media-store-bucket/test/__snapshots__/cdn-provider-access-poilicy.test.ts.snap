// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CDN provider access policy construct for the media bucket - it should create a cdn provider access policy without a default IAM user. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testbucketE6E05ABE": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testcdnprovideraccesspolicyAccessPolicyACF8C570": {
      "Properties": {
        "Groups": [
          {
            "Ref": "testcdnprovideraccesspolicyGroupBE650587",
          },
        ],
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:ListBucket",
                "s3:GetBucketLocation",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testbucketE6E05ABE",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "testbucketE6E05ABE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcdnprovideraccesspolicyAccessPolicyACF8C570",
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcdnprovideraccesspolicyGroupBE650587": {
      "Type": "AWS::IAM::Group",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`CDN provider access policy construct for the media bucket - it should create a default cdn provider access policy. 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testbucketE6E05ABE": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "testcdnprovideraccesspolicyAccessPolicyACF8C570": {
      "Properties": {
        "Groups": [
          {
            "Ref": "testcdnprovideraccesspolicyGroupBE650587",
          },
        ],
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:ListBucket",
                "s3:GetBucketLocation",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "testbucketE6E05ABE",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "testbucketE6E05ABE",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testcdnprovideraccesspolicyAccessPolicyACF8C570",
      },
      "Type": "AWS::IAM::Policy",
    },
    "testcdnprovideraccesspolicyGroupBE650587": {
      "Type": "AWS::IAM::Group",
    },
    "testcdnprovideraccesspolicyUserA4A3C503": {
      "Properties": {
        "Groups": [
          {
            "Ref": "testcdnprovideraccesspolicyGroupBE650587",
          },
        ],
      },
      "Type": "AWS::IAM::User",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
